dali/pipeline/executor/async_pipelined_executor.h

```diff
@@ -42,13 +42,24 @@ class DLL_PUBLIC AsyncPipelinedExecutor : public PipelinedExecutor {
         cpu_thread_(device_id, set_affinity),
         mixed_thread_(device_id, set_affinity),
         gpu_thread_(device_id, set_affinity),
-        device_id_(device_id) {}
+        device_id_(device_id),
+        should_stop_(false) {}
 
   DLL_PUBLIC ~AsyncPipelinedExecutor() override {
     ShutdownQueue();
     cpu_thread_.ForceStop();
     mixed_thread_.ForceStop();
     gpu_thread_.ForceStop();
+
+    /*
+     * We need to set should_stop_ and notify all work that may is scheduled that it
+     * should stop now. It may happen that mixed and GPU stages are scheduled, but only GPU
+     * one is picked up while mixed is discarded as the worker thread is already shutting down.
+     * In the end, GPU state will wait infinitely for mixed one. This code defends against it.
+     */
+    should_stop_ = true;
```

Comment:
You can use `IsStopSignaled()` instead of this variable - it's already set by the call to `ShutdownQueue`.